### ISSUE TITLE
Error: Duplicate list element ID when trying to insert at the same position

### ISSUE THREAD

**Issue by minhhien1996** [automerge/automerge#151]
I am trying to use Automerge for SlateJS. Then I when I am trying to make some changes by splitting a value whose type is `list` at then same position twice then the error thrown.

Here are steps to reproduce:

- Init
```const Automerge = require('automerge');

const empty = Automerge.init();

const init = `{
  "object": "value",
  "document": {
    "object": "document",
    "data": {
      "order": 3
    },
    "nodes": [
      {
        "object": "block",
        "data": {},
        "isVoid": false,
        "nodes": [
          {
            "object": "text",
            "leaves": [
              {
                "object": "leaf",
                "marks": [],
                "text": [
                  "a",
                  "a",
                  "a"
                ]
              }
            ]
          }
        ],
        "type": "paragraph"
      }
    ]
  }
}`;

const assertPath = (docRoot, path) => {
  let currentNode = docRoot;
  path.forEach(el => {
    currentNode = currentNode.nodes[el];
  });
  return currentNode;
};

const clone = x =>  JSON.parse(JSON.stringify(x));

let doc = Automerge.change(empty, '1', doc => {
  doc.note = JSON.parse(init);
});
```

- My handler for the split
```
const handleSplit = ({
  position, rest, index,
  currentNode, isText,
}) => {
  if (isText) {
    currentNode = assertPath(currentNode, rest);
    nodeToSplit = currentNode.nodes[index];
    cloned = clone(nodeToSplit);

    nodeToSplit.leaves[0].text.splice(position);
    cloned.leaves[0].text.splice(0, position);

    console.log('insert to path', [...rest, index + 1])
  } else {
    currentNode = assertPath(currentNode, rest);
    nodeToSplit = currentNode.nodes[index];
    cloned = clone(nodeToSplit);
  
    nodeToSplit.nodes.splice(position);
    cloned.nodes.splice(0, position);
  
    console.log('insert to path', [...rest, index + 1])
  }

  // currentNode.nodes.splice(index + 1, 0, cloned);
  currentNode.nodes.insertAt(index + 1, cloned);
}
```
- Split at path `doc.note.document.nodes.0.leaves.0` at position 2, then split at path `doc.note.document.nodes.0.leaves` at position 1. In short, I am trying to create a new element by cloning the element in the list and then insert it back to the list. 2 operations here is inherited from Slate.

```
doc = Automerge.change(doc, 'split', doc => {
  let rest, position, index, currentNode;
  //  [0, 0] position 2
  position = 2
  rest = [0]
  index = 0

  currentNode = doc.note.document;
  handleSplit({
    position, rest, index,
    currentNode, isText: true,
  })

  //  [0] position 1
  position = 1
  rest = []
  index = 0

  currentNode = doc.note.document;
  handleSplit({
    position, rest, index,
    currentNode, isText: false,
  })
});
```

- the result after these steps should be

```
{
  "object": "value",
  "document": {
    "object": "document",
    "data": {
      "order": 3
    },
    "nodes": [
      {
        "object": "block",
        "data": {},
        "isVoid": false,
        "nodes": [
          {
            "object": "text",
            "leaves": [
              {
                "object": "leaf",
                "marks": [],
                "text": [
                  "a",
                  "a"
                ]
              }
            ]
          }
        ],
        "type": "paragraph"
      },
      {
        "object": "block",
        "data": {},
        "isVoid": false,
        "nodes": [
          {
            "object": "text",
            "leaves": [
              {
                "object": "leaf",
                "marks": [],
                "text": [
                  "a"
                ]
              }
            ]
          }
        ],
        "type": "paragraph"
      }
    ]
  }
}
```

- Split at path `doc.note.document.nodes.0.leaves.0` at position 1 then split at path `doc.note.document.nodes.0.leaves` at position 1:

```doc = Automerge.change(doc, 'split', doc => {
  let rest, position, index, currentNode;
  //  [0, 0] position 1
  position = 1
  rest = [0]
  index = 0

  currentNode = doc.note.document;
  handleSplit({
    position, rest, index,
    currentNode, isText: true,
  })

  //  [0] position 1
  position = 1
  rest = []
  index = 0

  currentNode = doc.note.document;
  handleSplit({
    position, rest, index,
    currentNode, isText: false,
  })
});
```

- then the error is thrown:

```
Error: Duplicate list element ID 62092e48-5bd0-495a-b762-ac32c728baff:2
    at applyInsert (/Users/minhhien1996/Projects/hasBrain/automerge-server/node_modules/automerge/dist/automerge.js:795:72)
    at applyOps (/Users/minhhien1996/Projects/hasBrain/automerge-server/node_modules/automerge/dist/automerge.js:1049:28)
    at applyChange (/Users/minhhien1996/Projects/hasBrain/automerge-server/node_modules/automerge/dist/automerge.js:1104:19)
    at applyQueuedOps (/Users/minhhien1996/Projects/hasBrain/automerge-server/node_modules/automerge/dist/automerge.js:1137:30)
    at Object.addChange (/Users/minhhien1996/Projects/hasBrain/automerge-server/node_modules/automerge/dist/automerge.js:1190:27)
    at apply (/Users/minhhien1996/Projects/hasBrain/automerge-server/node_modules/automerge/dist/automerge.js:404:36)
    at Object.applyLocalChange (/Users/minhhien1996/Projects/hasBrain/automerge-server/node_modules/automerge/dist/automerge.js:461:18)
    at makeChange (/Users/minhhien1996/Projects/hasBrain/automerge-server/node_modules/automerge/dist/automerge.js:3385:54)
    at Object.change (/Users/minhhien1996/Projects/hasBrain/automerge-server/node_modules/automerge/dist/automerge.js:3599:12)
    at Object.change (/Users/minhhien1996/Projects/hasBrain/automerge-server/node_modules/automerge/dist/automerge.js:14004:35)
```

**Comment by minhhien1996** [comment]
I try both `insertAt` and `splice` with the element to be inserted as the third parameter, but non of them working

**Comment by minhhien1996** [comment]
To make my case not crashed, I change the function `applyInsert` in the `Backend.op_set`:

```
if (opSet.hasIn(['byObject', objectId, '_insertion', elemId])) throw new Error('Duplicate list element ID ' + elemId)
```

into 

```
if (opSet.hasIn(['byObject', objectId, '_insertion', elemId])) {
    const oldOp = opSet.getIn(['byObject', objectId, '_insertion', elemId])
    const oldSeq = oldOp.get('seq');
    const newSeq = op.get('seq');
    const isOpsConcurrent = isConcurrent(opSet, op, oldOp);
    if (isOpsConcurrent || oldSeq > newSeq) {
      throw new Error('Duplicate list element ID ' + elemId)
    }
  }
```

**Comment by lovegnep** [comment]
I also seen the problem

**Comment by ept** [comment]
@minhhien1996 Thank you for your detailed bug report! I was able to reproduce the issue using your code, and implemented a fix on #153. Could you please give it a try and see whether it fixes the problem for you?

(I'm afraid the change you propose to `applyInsert` is not the right solution.)
